src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,419 @@
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  data::NormalizeLabels(rawLabels, labels, mappings);
+
+  // Load the model, if necessary.
+  LinearSVM<>* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVM<>*>("input_model");
+  }
+  else
+  {
+    numClasses = CLI::GetParam<int>("number_of_classes") == 0 ?
+        arma::max(labels) + 1 : CLI::GetParam<int>("number_of_classes");
+    model = new LinearSVM<>(0, numClasses);
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    model->Lambda() = lambda;
+    model->Delta() = delta;
+    numClasses = model->NumClasses();
+
+    if (optimizerType == "lbfgs")
+    {
+      model->FitIntercept() = intercept;
+
+      ens::L_BFGS lbfgsOpt;
+      lbfgsOpt.MaxIterations() = maxIterations;
+      lbfgsOpt.MinGradientNorm() = tolerance;
+
+      Log::Info << "Training model with L-BFGS optimizer." << endl;
+
+      // This will train the model.
+      model->Train(trainingSet, labels, numClasses, lbfgsOpt);
+    }
+    else if (optimizerType == "psgd")
+    {
+      const double step_size = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(step_size);
+
+      #ifndef HAS_OPENMP
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, trainingSet.n_cols,
+        tolerance, shuffle, decayPolicy);
+      #endif
+
+      // This optimizer is only compiled if OpenMP is used.
+      #ifdef HAS_OPENMP
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, std::ceil(
+        (float)trainingSet.n_cols / omp_get_max_threads()), tolerance, shuffle,
+        decayPolicy);
+      #endif
+
+      Log::Info << "Training model with ParallelSGD optimizer." << endl;
+
+      // This will train the model.
+      model->Train(trainingSet, labels, numClasses, psgdOpt);
+    }
+  }
+  if (CLI::HasParam("test"))
+  {
+    if (!CLI::HasParam("training"))
+    {
+      numClasses = model->NumClasses();
+    }
+    // Get the test dataset, and get predictions.
+    testSet = std::move(CLI::GetParam<arma::mat>("test"));
+    arma::Row<size_t> predictions;
+    size_t trainingDimensionality;
+
+    // Set the dimensionality according to fitintercept.
+    if (intercept && optimizerType == "lbfgs")
+      trainingDimensionality = model->Parameters().n_rows - 1;
+    else
+      trainingDimensionality = model->Parameters().n_rows;
+
+    // Checking the dimensionality of the test data.
+    if (testSet.n_rows != trainingDimensionality)
+    {
+      Log::Fatal << "Test data dimensionality (" << testSet.n_rows << ") must "
+          << "be the same as the dimensionality of the training data ("
+          << trainingDimensionality << ")!" << endl;
+    }
+
+    // Save class score, if desired.
+    if (CLI::HasParam("score"))
+    {
+      Log::Info << "Calculating class score of points in '"
```

Comment:
No need for a `'` here around the output of `GetPrintableParam()`.  In the cases where a filename is printed, I think it is already enclosed in `''`s, and when it isn't, it just prints the size of the matrix, which looks a bit odd in quotes. :)

Reply:
Removed extra `'`. :)